### PR TITLE
Data access hook changes

### PR DESCRIPTION
**Pull request by samcook**
A few changes to Data Access Hooks:

- adds `BeforeRollback` and `AfterRollback` hooks
- adds IDs to `DataAccessTransaction` and `TransactionContext` which are available in the Submit and Rollback hooks
- Adds some tests around submit and rollback hooks
- Fixes some bugs

### CODE REVIEW

File: src/Shaolinq/TransactionContext.cs [hunk 1]
```diff
@@ -146,6 +148,8 @@ private static TransactionContext GetOrCreateCurrent(DataAccessModel dataAccessM
 
 				return context;
 			}
+
+			dataAccessTransaction.CheckAborted();
```
Comment (samcook) [inline src/Shaolinq/TransactionContext.cs]: Without this check you end up hitting an `ObjectDisposedException` when using `DataAccessScope` if you try and continue using an outer scope after rolling back an inner scope. This makes it consistent with `TransactionScope`, which throws a `TransactionAbortedException`.

File: tests/Shaolinq.Tests/DataAccessModelHookTests.cs [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+				{
+					var cat3 = this.model.Cats.Create();
+				});
+			}
+		}
+
+
+		[TestCase(false, false)]
+		[TestCase(false, true)]
+		[TestCase(true, false, Ignore = "Hangs")]
```
Comment (samcook) [inline tests/Shaolinq.Tests/DataAccessModelHookTests.cs]: This test is a bit of an edge case, but this combination causes the test to hang completely.
